### PR TITLE
Add outer try block for obtainKey errors. Add docs.

### PR DESCRIPTION
**Pull request by iskakaushik**
- Added some documentation around the error handling in `ImageProvider`.
- Added additional error handling to catch errors in `ImageProvide.obtainKey` and added a test for the same.

### CODE REVIEW

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: Use `//` for comments and `///` for dart docs.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: Instead of wrapping everything in a try-catch, just wrap the invocation of obtainKey.

```
Future<T> obtainedKey;
try {
  obtainedKey = obtainKey(configuration);
}
```

Comment (iskakaushik) [reply packages/flutter/lib/src/painting/image_provider.dart]: There was a variable `obtainedKey` already in the scope, so i used `key` instead.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: If you catch an error here `key` will be null, right?

Comment (iskakaushik) [reply packages/flutter/lib/src/painting/image_provider.dart]: 🤕 yeah, sorry will fix this

Comment (jonahwilliams) [reply packages/flutter/lib/src/painting/image_provider.dart]: Please add a test case which exercises this new code

**Comment by jonahwilliams** [comment]
Please also leave a description on the PR so that future archeologists can better understand this change

File: packages/flutter/test/painting/image_provider_test.dart
Comment (jonahwilliams) [inline packages/flutter/test/painting/image_provider_test.dart]: Can you confirm that this test fails without your change?

Comment (iskakaushik) [reply packages/flutter/test/painting/image_provider_test.dart]: Yup.
```
~/C/f/f/p/flutter ❯❯❯ flutter test ./test/painting/image_provider_test.dart                                                                                                                                                       ✘ 1 
Running "flutter packages get" in flutter...                        0.5s
00:04 +4 -1: ImageProvide.obtainKey errors will be caught [E]                                                                                                                                                                         
  Instance of 'Error'
  test/painting/mocks_for_image_cache.dart 105:5          ObtainKeyErrorImageProvider.obtainKey
  package:flutter/src/painting/image_provider.dart 283:5  ImageProvider.resolve
  test/painting/image_provider_test.dart 81:46            main.<fn>
  ===== asynchronous gap ===========================
  dart:async                                              _AsyncAwaitCompleter.completeError
  test/painting/image_provider_test.dart                  main.<fn>
  
00:04 +4 -1: Some tests failed.     
```

**Review by jonahwilliams — APPROVED** [review]
LGTM

please wait until the tree is open to land.

**Pull request merged**
